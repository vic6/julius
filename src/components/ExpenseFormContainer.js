import React, { Component } from 'react';
import { Form, Message } from 'semantic-ui-react';
import ExpenseForm from './ExpenseForm';

export default class ExpenseFormContainer extends Component {
  state = {
    expenseName: '',
    amount: '',
    payer: {},
    consumers: [],
    errors: false,
    success: null
  };

  handleChange = event => {
    const { name, value } = event.target;
    this.setState({
      [name]: value
    });
  };

  goBackExpenseForm = history => {
    history.push('/event');
  };

  handleAmountChange = event => {
    const amount = event.target.value;
    const currencyCheck = /^\d{1,}(\.\d{0,2})?$/;
    // allows user to clear amount
    if (!amount || amount.match(currencyCheck)) {
      this.setState(() => ({ amount }));
    }
  };

  resetExpenseForm = () => {
    this.setState({
      payer: '',
      expenseName: '',
      amount: '',
      consumers: []
    });
  };

  handleSelectChange = event => {
    const participants = JSON.parse(localStorage.getItem('participants'));
    const payer =
      participants.filter(participant => participant.name === event.target.innerText)[0] || '';
    this.setState({ payer });
  };

  toggleCheckbox = particpant => {
    if (this.state.consumers.includes(particpant.name)) {
      this.setState({
        consumers: this.state.consumers.filter(person => person !== particpant.name)
      });
    } else {
      this.setState({
        consumers: [...this.state.consumers, particpant.name]
      });
    }
  };

  isFormValid = () => {
    if (this.state.payer.name === undefined) {
      this.setState({
        errors: 'Please select a payer'
      });
      return false;
    } else if (this.state.consumers.length < 1) {
      this.setState({ errors: 'Select at least 1 person to charge' });
      return false;
    }
    this.setState({
      errors: false,
      success: <Message success header={`Expense Added: ${this.state.expenseName}`} />
    });
    return true;
  };

  renderCheckboxes = () =>
    JSON.parse(localStorage.getItem('participants')).map(person => (
      <Form.Field
        key={person.name}
        onChange={event => this.toggleCheckbox(person, event)}
        label={person.name}
        value={person.name}
        control="input"
        type="checkbox"
      />
    ));

  render() {
    const participants = JSON.parse(localStorage.getItem('participants'));
    const options = participants.map(person => ({
      text: person.name,
      value: person.name
    }));
    const { expenseName, amount, payer, consumers, errors, success } = this.state;
    return (
      <ExpenseForm
        handleAddExpense={this.props.handleAddExpense}
        handleChange={this.handleChange}
        handleAmountChange={this.handleAmountChange}
        resetExpenseForm={this.resetExpenseForm}
        handleSelectChange={this.handleSelectChange}
        toggleCheckbox={this.toggleCheckbox}
        isFormValid={this.isFormValid}
        renderCheckboxes={this.renderCheckboxes}
        options={options}
        expenseName={expenseName}
        amount={amount}
        payer={payer}
        consumers={consumers}
        errors={errors}
        success={success}
        history={this.props.history}
        goBackExpenseForm={this.goBackExpenseForm}
      />
    );
  }
}
