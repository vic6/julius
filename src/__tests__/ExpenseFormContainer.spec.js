import React from 'react';
import renderer from 'react-test-renderer';
import ExpenseFormContainer from '../components/ExpenseFormContainer';

test('ExpenseFormContainer renders correctly', ()=> {
  const participants = [{name: 'chef', profile: {}}, {name: 'bob', profile: {}}, {name: 'stan', profile: {}}]
  const expenses = {expenseName: 'Girl scout cookies', amount: '3.50', payer: participants[0], consumers: [participants[1], participants[2]]}

  localStorage.setItem('participants', JSON.stringify(participants))
  localStorage.setItem('expenses', JSON.stringify(expenses))

  const component = renderer.create(<ExpenseFormContainer />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})
