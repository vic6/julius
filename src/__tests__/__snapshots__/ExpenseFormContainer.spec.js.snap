// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseFormContainer renders correctly 1`] = `
<div>
  <form
    action={undefined}
    autoComplete="off"
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="required field"
    >
      <label
        htmlFor={undefined}
      >
        Who Paid
      </label>
      <div
        aria-busy={undefined}
        aria-disabled={undefined}
        aria-expanded={false}
        aria-multiselectable={undefined}
        className="ui selection dropdown"
        content={undefined}
        name="payer"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        required={true}
        role="listbox"
        tabIndex={0}
        type={undefined}
      >
        <div
          aria-live="polite"
          className="default text"
          role="alert"
        >
          Name
        </div>
        <i
          aria-hidden="true"
          className="dropdown icon"
          onClick={[Function]}
        />
        <div
          className="menu transition"
        >
          <div
            aria-checked={false}
            aria-disabled={undefined}
            aria-selected={true}
            className="selected item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              chef
            </span>
          </div>
          <div
            aria-checked={false}
            aria-disabled={undefined}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              bob
            </span>
          </div>
          <div
            aria-checked={false}
            aria-disabled={undefined}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              stan
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="required field"
    >
      <label
        htmlFor={undefined}
      >
        Item name
      </label>
      <div
        className="ui input"
        content={undefined}
      >
        <input
          disabled={undefined}
          name="expenseName"
          onChange={[Function]}
          placeholder="flashlight"
          required={true}
          tabIndex={undefined}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="required field"
    >
      <label
        htmlFor={undefined}
      >
        Amount
      </label>
      <div
        className="ui input"
        content={undefined}
      >
        <input
          disabled={undefined}
          name="amount"
          onChange={[Function]}
          required={true}
          tabIndex={undefined}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
      name="consumers"
    >
      <label>
        Split Between
      </label>
    </div>
    <div
      className="field"
    >
      <label>
        <input
          content={undefined}
          disabled={undefined}
          onChange={[Function]}
          required={undefined}
          type="checkbox"
          value="chef"
        />
         
        chef
      </label>
    </div>
    <div
      className="field"
    >
      <label>
        <input
          content={undefined}
          disabled={undefined}
          onChange={[Function]}
          required={undefined}
          type="checkbox"
          value="bob"
        />
         
        bob
      </label>
    </div>
    <div
      className="field"
    >
      <label>
        <input
          content={undefined}
          disabled={undefined}
          onChange={[Function]}
          required={undefined}
          type="checkbox"
          value="stan"
        />
         
        stan
      </label>
    </div>
    <button
      className="ui primary button"
      disabled={undefined}
      onClick={[Function]}
      role="button"
      tabIndex={undefined}
    >
      Enter Item
    </button>
    <button
      className="ui secondary button"
      disabled={undefined}
      onClick={[Function]}
      role="button"
      tabIndex={undefined}
    >
      Back
    </button>
  </form>
</div>
`;
